Add audit.open_api_auth plugin

artem-smotrakov opened the pull request:
This pull request introduces a new plugin `audit.open_api_auth` which is described in #17343 

Implementation notes:

- The new plugin works only with API endpoints. It requires an API specification which has to be provided by `crawl.open_api` plugin.
- The `crawl.open_api` plugin shares API specification via the knowledge base.
- The new plugin simply checks that endpoints return 401 error if no authentication info is provided in the request.

Please consider it as a first version of the plugin. I was going to make the plugin smaller but it turned out that it needs quite a lot of checks, so now the plugin is not that small :) 

I believe the plugin may be improved. Please take a look at several enhancements in the plugin description, and let me know what you think. I am planning to work on them but would prefer to implement them separately after the first version of the plugin is integrated.

artem-smotrakov committed:
Update open_api to store a copy of SpecificationHandler in the kb

artem-smotrakov committed:
Added first draft of open_api_auth.py

artem-smotrakov committed:
Updated open_api_auth plugin to check and remove auth info from requests

artem-smotrakov committed:
Added test_open_api_auth.py and fixed issues in open_api_auth

artem-smotrakov committed:
open_api_auth should take into account 'security' sections of operations

artem-smotrakov committed:
Updated mock responses in TestOpenAPIAuth

artem-smotrakov committed:
NewPet and Error should be objects in petstore_with_security.yaml

artem-smotrakov committed:
open_api_auth should take into account 'security' sections of operations for all auth schemes

artem-smotrakov committed:
Check response codes in open_api_auth plugin

artem-smotrakov committed:
Split tests in test_open_api_auth.py

artem-smotrakov committed:
Describe open_api_auth plugin

artem-smotrakov committed:
Describe methods in open_api_auth plugin

artem-smotrakov committed:
Describe new methods in OpenAPI class

artem-smotrakov committed:
Describe new methods in SpecificationHandler

andresriancho commented on w3af/plugins/audit/open_api_auth.py:
```diff
@@ -0,0 +1,372 @@
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self._spec = None
+        self._expected_codes = [401]
```
Maybe add 403?

artem-smotrakov replied:
I also thought about it. In a perfect world, an application should return 401 if no auth info is present at all. 403 should be usually returned if a request contains auth info, but authorization failed. So now the plugin assumes that it lives in a perfect world :) I think it is not a big deal If an application returns 403 instead of 401 but strictly speaking it doesn't sound correct to me. Anyway, I don't mind adding 403 here. I would even let a user to configure it - it may be next improvement.

andresriancho commented on w3af/plugins/audit/open_api_auth.py:
```diff
@@ -0,0 +1,372 @@
+        # The check works only with API endpoints.
+        # crawl.open_api plugin has to be called before.
+        if not self._is_api_spec_available():
+            om.out.information("Could not find an API specification, skip")
```
During a regular w3af run, this line will be printed to the output MANY times. I recommend using `om.out.debug` here.

andresriancho replied:
Also the message could be improved: "Could not find an API specification in the KB, skipping open_api_auth"

andresriancho commented on w3af/plugins/audit/open_api_auth.py:
```diff
@@ -0,0 +1,372 @@
+
+        # The API spec must define authentication mechanisms.
+        if not self._has_security_definitions_in_spec():
+            om.out.information("The API spec has no security definitions, skip")
```
Change message to `The Open API spec has no security definitions, not running open_api_auth checks.`

Also, this should be printed only once per each open-api specification document w3af finds.

andresriancho commented on w3af/plugins/audit/open_api_auth.py:
```diff
@@ -0,0 +1,372 @@
+        # The check only runs if the API specification
+        # requires authentication for a endpoint.
+        if not self._has_auth(freq):
+            om.out.information("Request doesn't have auth info, skip")
```
Improve message and change to debug.

andresriancho commented on w3af/plugins/audit/open_api_auth.py:
```diff
@@ -0,0 +1,372 @@
+        :param response: The HTTP response associated with the fuzzable request.
+        """
+        if response.get_code() not in self._expected_codes:
+            desc = 'A %s request without auth info was send to %s. ' \
```
`... request without authentication information ...`

`was sent to`

`replied with unexpected HTTP response code %d (expected one of 401, 403, etc.)`

And add: `This is a strong indicator of a REST API authentication bypass`.

andresriancho commented on w3af/plugins/audit/open_api_auth.py:
```diff
@@ -0,0 +1,372 @@
+                      response.get_code(), self._expected_codes)
+
+            # Should severity depend on response codes?
+            # For example, 2xx codes results to HIGH, but 4xx may be MEDIUM/LOW
```
Yes, this is a good idea to remove false positives.

The way I recommend doing it is to add a new check below `if not self._has_auth(freq)`, like this:

```python
        if not self._response_is_successful(freq):
            om.out.debug("OpenAPI request {some request info here} did not return a successful response. Skipping open_api_auth check for this endpoint.")
```

Where `_response_is_successful` would send the request (with credentials) and check that the response code is 200, 201, or some other codes which are commonly used to indicate success.

artem-smotrakov replied:
Do you mean adding `_response_is_successful` check between `_remove_auth` and `_remove_auth` and `_uri_opener.send_mutant` calls in `audit()` method? But wasn't this request already sent? Is the result already available in `orig_response`?

What would be a definition of a successful response? 200 may be not enough since a endpoint may return 201, 202, etc. In fact, response codes should be defined in the spec. In general, it would be a nice improvement. Do you think such a check is strictly necessary?

In fact, I meant another thing here. I am talking about assigning different severity levels what we report a vulnerability to the KB. It should be done if/after `_response_is_successful` check passed.

andresriancho commented on w3af/plugins/audit/open_api_auth.py:
```diff
@@ -0,0 +1,372 @@
+        if self._spec:
+            return True
+
+        specification_handler = kb.kb.raw_read('open_api', 'specification_handler')
```
What if there is more than one openapi document found during a scan?

Maybe you should store a list and call it `kb.kb.raw_read('open_api', 'specification_handlers')` (added an `s` at the end)

This also means that you'll have to touch other areas of this plugin to process all specification handlers, not just one.

andresriancho commented on w3af/plugins/audit/open_api_auth.py:
```diff
@@ -0,0 +1,372 @@
+          * Check if the response has 401 error code (access denied).
+
+        A couple of important notes:
+          * The plugin requires the 'crawl.open_api' plugin to be enabled.
```
You can force that. Plugins can have other plugins as dependencies.
https://github.com/andresriancho/w3af/blob/master/w3af/core/controllers/plugins/plugin.py#L123

andresriancho replied:
When plugin A depends on B, and plugin A is enabled, plugin B will be automatically enabled too.

artem-smotrakov replied:
That's nice, I didn't know about it. Let me add such a dependency.

andresriancho reviewed (commented):
Awesome job!

artem-smotrakov committed:
Merge remote-tracking branch 'upstream/develop' into openapi-auth-plugin

artem-smotrakov committed:
Better checks for API spec in audit.open_api_auth plugin

artem-smotrakov committed:
Don't use _spec_is_good flag in audit.open_api_auth plugin

artem-smotrakov committed:
audit.open_api_auth should depend on crawl.open_api plugin

artem-smotrakov committed:
audit.open_api_auth plugin should expect multiple API specs

artem-smotrakov committed:
Added TestOpenAPIAuthMultipleSpecs

artem-smotrakov commented:
Here is a couple of updates:

- `audit.open_api_auth` now depends on `crawl.open_api`.
- `audit.open_api_auth` now expects multiple specs.
- Updated debug messages.
- Added `TestOpenAPIAuthMultipleSpecs`.

I also have a couple of comments/questions. Please let me know what you think.

artem-smotrakov commented:
Still waiting for feedback (and for other pull requests too). Please take a look when you have time.
